Add a non body consuming check for form boundaries

jon-betts opened the pull request:
This also adds an option to control whether blank content types should be
assumed to be a form type for the purpose of checking when getting a form
for the purposes of this check.

The default is False, as otherwise installing `pyramid-sanity` could cause
otherwise working apps to fail (e.g. ones expecting JSON, but not requiring
a content type to declare it).

jon-betts commented on README.md:
```diff
@@ -24,24 +24,12 @@ with Configurator() as config:
-
-    # Don't safely encode redirect locations.
-    "pyramid_sanity.ascii_safe_redirects": False
+    "pyramid_sanity.check_form": False
```
I shifted this stuff about for a few reasons:

 * This section is about disabling things, it didn't make sense to list the `pyramid_sanity.check_form.assume_form_on_blank` here
* This is more usage than documentation of the settings. It was trying to do double duty, but I've split them up

jon-betts commented on README.md:
```diff
@@ -54,10 +42,22 @@ config.add_settings({
+| `pyramid_sanity.check_form.assume_form_on_blank` | `False` | Assume it's a form when the Content-Type is blank. This can cause errors when accepting non-form data
+| `pyramid_sanity.check_params` | `True` | Check for badly encoded query params
+| `pyramid_sanity.check_path` | `True` | Check for badly encoded URL paths
+| `pyramid_sanity.ascii_safe_redirects` | `True` | Safely encode redirect locations
```
Positively state the effect of each option (rather than "Don't...")

jon-betts commented on src/pyramid_sanity/tweens.py:
```diff
@@ -41,21 +44,51 @@ def tween(request):
 
 @tween_factory
-def invalid_form_tween_factory(request, handler, _registry):
+def invalid_form_tween_factory(request, handler, registry):
```
Turns out adding `registry` was precient!

jon-betts commented on src/pyramid_sanity/tweens.py:
```diff
@@ -41,21 +44,51 @@ def tween(request):
     return handler(request)
 
 
+def _check_form_boundary(request, assume_form=False):
```
Use an inner function to allow quick returns, rather than nested ifs

seanh commented:
Can you explain the `assume_form_on_blank` setting? It looks like previously the invalid form tween did its thing on all `POST` requests. Now, by default, it only does its thing if the request method is `POST` and the `Content-Type` is `"application/x-www-form-urlencoded"` or `"multipart/form-data"`. 

Why the change? It looks to me like this is to fix the bulk API because, since bulk API requests aren't `"application/x-www-form-urlencoded"` or `"multipart/form-data"`, the invalid form tween has now been disabled for bulk API requests? So the `request.POST` iterable will no longer be getting consumed by the tween, for bulk API tests?

Why the config setting to undo the change? Wouldn't it make sense to _always_ limit the invalid form tween to `"application/x-www-form-urlencoded"` or `"multipart/form-data"` requests?

seanh commented:
Hmm, and it seems that `request.POST.get("", None)` was replaced with  cgi.valid_boundary()`. I'm not sure what `valid_boundary()` does though, it's undocumented!

jon-betts commented:
It's "inspired" rather than directly related. It's not needed to fix the bulk API issue per se, but will affect 3rd parties who try to use `pyramid-sanity`. Consider the following situation:

 * You setup a POST view, that is expecting JSON for example
 * You do not require any particular `Content-Type` header (so it's fine to send anything, or nothing at all)
 * You therefore end-up with consumers (self or other) that make requests, sending JSON, without a `Content-Type`
 * Now you enable `pyramid-sanity`
 * As a result, previously working end-points are broken raising messages about bad form boundaries, which is wrong and hyper confusing

I can't see a sane way of proceeding if the `Content-Type` is blank that doesn't break people. If you say it's a form and it's broken, than seems fair enough. If you don't even claim it to be a form, it's a bit weird. However by removing this, we strip protection for that. So I wanted a way to add that protection back in if you wanted it.

This is pretty much the situation we are actually in with `h` and the bulk API. We don't enforce that you tell us the content type to be declared, we just assume it's NDJSON. This meant that although the client sends the correct header, the func tests weren't. It turns out `webtest` tries to be helpful, and defaults to "application/x-www-form-urlencoded" so they crashed.

So while the actual type of the content we are sending is NDJSON for the bulk API, the `Content-Type` can be anything, or nothing at all.

jon-betts commented:
The `cgi` stuff is stripped out of the `webob.compat` stuff mentioned in the code.

jon-betts commented:
This has undergone some fairly major changes:

 * I added full testing of the error condition and realised it only applied to "multipart" things (which makes sense)
 * This removes the need for the option
 * I added testing of the others to catch similar things
 * Added suggested wording for docstring above
